src/Conversion/ONNXToKrnl/Tensor/ConcatShapeTranspose.cpp

```diff
@@ -0,0 +1,182 @@
+
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===-------------- Concat.cpp - Lowering ConcatShapeTranspose Op ---------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Concat OperatorShapeTranspose to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/Krnl/KrnlHelper.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+struct ONNXConcatShapeTransposeOpLowering : public ConversionPattern {
+  ONNXConcatShapeTransposeOpLowering(
+      TypeConverter &typeConverter, MLIRContext *ctx)
+      : ConversionPattern(typeConverter,
+            mlir::ONNXConcatShapeTransposeOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    auto loc = op->getLoc();
+
+    ONNXConcatShapeTransposeOpAdaptor operandAdaptor(
+        operands, op->getAttrDictionary());
+    MultiDialectBuilder<KrnlBuilder, MathBuilder> create(rewriter, loc);
+
+    // Compute concat output shape.
+    unsigned numInputs = op->getNumOperands();
+    Value firstInput = operandAdaptor.inputs().front();
+    ArrayRef<int64_t> commonShape =
+        firstInput.getType().cast<ShapedType>().getShape();
+    // Type dataElementType =
+    // firstInput.getType().cast<ShapedType>().getElementType();
+    uint64_t commonRank = commonShape.size();
+    int64_t axisIndex = operandAdaptor.axis();
+
+    // Negative axis means values are counted from the opposite side.
+    // TOFIX should be in normalization pass
+    if (axisIndex < 0)
+      axisIndex += commonRank;
+
+    IndexExprScope IEScope(&rewriter, loc);
+    DimsExpr outputConcatDims(commonRank);
```

Comment:
Highly recommend to create a ShapeHelp for this new operation and move the code here into it, which helps DimAnalysis to support it seamlessly.